File: sdk/python/aragora/namespaces/connectors.py
"""
Connectors Namespace API

Provides methods for managing enterprise data source connectors,
sync operations, and scheduler configuration.

Features:
- Connector management (GitHub Enterprise, S3, PostgreSQL, MongoDB, FHIR)
- Sync scheduling and monitoring
- Connector health checks
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Any, Literal

if TYPE_CHECKING:
    from ..client import AragoraAsyncClient, AragoraClient


ConnectorType = Literal["github_enterprise", "s3", "postgresql", "mongodb", "fhir"]
SyncFrequency = Literal["hourly", "daily", "weekly", "manual"]


class ConnectorsAPI:
    """
    Synchronous Connectors API.

    Provides methods for managing enterprise data source connectors:
    - Connector CRUD operations
    - Sync scheduling
    - Health monitoring

    Example:
        >>> client = AragoraClient(base_url="https://api.aragora.ai", api_key="...")
        >>> connectors = client.connectors.list()
        >>> client.connectors.trigger_sync(connector_id="abc123")
    """

    def __init__(self, client: AragoraClient):
        self._client = client

    # ===========================================================================
    # Connector Management
    # ===========================================================================

    def list(
        self,
        connector_type: ConnectorType | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List configured connectors.

        Args:
            connector_type: Filter by connector type
            limit: Maximum number of results
            offset: Pagination offset

        Returns:
            Dict with connectors array and pagination info
        """
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if connector_type:
            params["type"] = connector_type

        return self._client.request("GET", "/api/v1/connectors", params=params)

    def get(self, connector_id: str) -> dict[str, Any]:
        """
        Get a connector by ID.

        Args:
            connector_id: Connector ID

        Returns:
            Dict with connector details
        """
        return self._client.request("GET", f"/api/v1/connectors/{connector_id}")

    def create(
        self,
        name: str,
        connector_type: ConnectorType,
        config: dict[str, Any],
        schedule: SyncFrequency = "daily",
    ) -> dict[str, Any]:
        """
        Create a new connector.

        Args:
            name: Connector name
            connector_type: Type of connector
            config: Connector-specific configuration
            schedule: Sync frequency

        Returns:
            Dict with created connector details
        """
        return self._client.request(
            "POST",
            "/api/v1/connectors",
            json={
                "name": name,
                "type": connector_type,
                "config": config,
                "schedule": schedule,
            },
        )

    def update(
        self,
        connector_id: str,
        name: str | None = None,
        config: dict[str, Any] | None = None,
        schedule: SyncFrequency | None = None,
        enabled: bool | None = None,
    ) -> dict[str, Any]:
        """
        Update a connector.

        Args:
            connector_id: Connector ID
            name: New name
            config: New configuration
            schedule: New sync frequency
            enabled: Enable/disable connector

        Returns:
            Dict with updated connector details
        """
        data: dict[str, Any] = {}
        if name is not None:
            data["name"] = name
        if config is not None:
            data["config"] = config
        if schedule is not None:
            data["schedule"] = schedule
        if enabled is not None:
            data["enabled"] = enabled

        return self._client.request("PATCH", f"/api/v1/connectors/{connector_id}", json=data)

    def delete(self, connector_id: str) -> dict[str, Any]:
        """
        Delete a connector.

        Args:
            connector_id: Connector ID

        Returns:
            Dict with success status
        """
        return self._client.request("DELETE", f"/api/v1/connectors/{connector_id}")

    # ===========================================================================
    # Sync Operations
    # ===========================================================================

    def trigger_sync(
        self,
        connector_id: str,
        full_sync: bool = False,
    ) -> dict[str, Any]:
        """
        Trigger a sync for a connector.

        Args:
            connector_id: Connector ID
            full_sync: Whether to do a full sync (vs incremental)

        Returns:
            Dict with sync_id and status
        """
        return self._client.request(
            "POST",
            f"/api/v1/connectors/{connector_id}/sync",
            json={"full_sync": full_sync},
        )

    def get_sync_status(self, connector_id: str, sync_id: str) -> dict[str, Any]:
        """
        Get status of a sync operation.

        Args:
            connector_id: Connector ID
            sync_id: Sync operation ID

        Returns:
            Dict with sync status and progress
        """
        return self._client.request("GET", f"/api/v1/connectors/{connector_id}/syncs/{sync_id}")

    def list_syncs(
        self,
        connector_id: str,
        limit: int = 20,
    ) -> dict[str, Any]:
        """
        List recent sync operations for a connector.

        Args:
            connector_id: Connector ID
            limit: Maximum number of results

        Returns:
            Dict with syncs array
        """
        return self._client.request(
            "GET",
            f"/api/v1/connectors/{connector_id}/syncs",
            params={"limit": limit},
        )

    def cancel_sync(self, connector_id: str, sync_id: str) -> dict[str, Any]:
        """
        Cancel a running sync operation.

        Args:
            connector_id: Connector ID
            sync_id: Sync operation ID

        Returns:
            Dict with success status
        """
        return self._client.request(
            "POST", f"/api/v1/connectors/{connector_id}/syncs/{sync_id}/cancel"
        )

    # ===========================================================================
    # Health and Monitoring
    # ===========================================================================

    def test_connection(self, connector_id: str) -> dict[str, Any]:
        """
        Test connectivity for a connector.

        Args:
            connector_id: Connector ID

        Returns:
            Dict with connection_ok and latency_ms
        """
        return self._client.request("POST", f"/api/v1/connectors/{connector_id}/test")

    def get_health(self, connector_id: str) -> dict[str, Any]:
        """
        Get health status of a connector.

        Args:
            connector_id: Connector ID

        Returns:
            Dict with health status, last_sync, and error info
        """
        return self._client.request("GET", f"/api/v1/connectors/{connector_id}/health")


class AsyncConnectorsAPI:
    """
    Asynchronous Connectors API.

    Example:
        >>> async with AragoraAsyncClient(base_url="https://api.aragora.ai") as client:
        ...     connectors = await client.connectors.list()
        ...     await client.connectors.trigger_sync(connector_id="abc123")
    """

    def __init__(self, client: AragoraAsyncClient):
        self._client = client

    # ===========================================================================
    # Connector Management
    # ===========================================================================

    async def list(
        self,
        connector_type: ConnectorType | None = None,
        limit: int = 50,
        offset: int = 0,
    ) -> dict[str, Any]:
        """List configured connectors."""
        params: dict[str, Any] = {"limit": limit, "offset": offset}
        if connector_type:
            params["type"] = connector_type

        return await self._client.request("GET", "/api/v1/connectors", params=params)

    async def get(self, connector_id: str) -> dict[str, Any]:
        """Get a connector by ID."""
        return await self._client.request("GET", f"/api/v1/connectors/{connector_id}")

    async def create(
        self,
        name: str,
        connector_type: ConnectorType,
        config: dict[str, Any],
        schedule: SyncFrequency = "daily",
    ) -> dict[str, Any]:
        """Create a new connector."""
        return await self._client.request(
            "POST",
            "/api/v1/connectors",
            json={
                "name": name,
                "type": connector_type,
                "config": config,
                "schedule": schedule,
            },
        )

    async def update(
        self,
        connector_id: str,
        name: str | None = None,
        config: dict[str, Any] | None = None,
        schedule: SyncFrequency | None = None,
        enabled: bool | None = None,
    ) -> dict[str, Any]:
        """Update a connector."""
        data: dict[str, Any] = {}
        if name is not None:
            data["name"] = name
        if config is not None:
            data["config"] = config
        if schedule is not None:
            data["schedule"] = schedule
        if enabled is not None:
            data["enabled"] = enabled

        return await self._client.request("PATCH", f"/api/v1/connectors/{connector_id}", json=data)

    async def delete(self, connector_id: str) -> dict[str, Any]:
        """Delete a connector."""
        return await self._client.request("DELETE", f"/api/v1/connectors/{connector_id}")

    # ===========================================================================
    # Sync Operations
    # ===========================================================================

    async def trigger_sync(
        self,
        connector_id: str,
        full_sync: bool = False,
    ) -> dict[str, Any]:
        """Trigger a sync for a connector."""
        return await self._client.request(
            "POST",
            f"/api/v1/connectors/{connector_id}/sync",
            json={"full_sync": full_sync},
        )

    async def get_sync_status(self, connector_id: str, sync_id: str) -> dict[str, Any]:
        """Get status of a sync operation."""
        return await self._client.request(
            "GET", f"/api/v1/connectors/{connector_id}/syncs/{sync_id}"
        )

    async def list_syncs(
        self,
        connector_id: str,
        limit: int = 20,
    ) -> dict[str, Any]:
        """List recent sync operations for a connector."""
        return await self._client.request(
            "GET",
            f"/api/v1/connectors/{connector_id}/syncs",
            params={"limit": limit},
        )

    async def cancel_sync(self, connector_id: str, sync_id: str) -> dict[str, Any]:
        """Cancel a running sync operation."""
        return await self._client.request(
            "POST", f"/api/v1/connectors/{connector_id}/syncs/{sync_id}/cancel"
        )

    # ===========================================================================
    # Health and Monitoring
    # ===========================================================================

    async def test_connection(self, connector_id: str) -> dict[str, Any]:
        """Test connectivity for a connector."""
        return await self._client.request("POST", f"/api/v1/connectors/{connector_id}/test")

    async def get_health(self, connector_id: str) -> dict[str, Any]:
        """Get health status of a connector."""
        return await self._client.request("GET", f"/api/v1/connectors/{connector_id}/health")

File: sdk/typescript/src/namespaces/connectors.ts
/**
 * Connectors Namespace API
 *
 * Provides methods for managing enterprise data source connectors,
 * sync operations, and scheduler configuration.
 *
 * Features:
 * - Connector management (GitHub Enterprise, S3, PostgreSQL, MongoDB, FHIR)
 * - Sync scheduling and monitoring
 * - Connector health checks
 */

/**
 * Supported connector types.
 */
export type ConnectorType = 'github_enterprise' | 's3' | 'postgresql' | 'mongodb' | 'fhir';

/**
 * Sync frequency options.
 */
export type SyncFrequency = 'hourly' | 'daily' | 'weekly' | 'manual';

/**
 * Sync operation status.
 */
export type SyncStatus = 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';

/**
 * Connector health status.
 */
export type HealthStatus = 'healthy' | 'degraded' | 'unhealthy' | 'unknown';

/**
 * Connector details.
 */
export interface Connector {
  id: string;
  name: string;
  type: ConnectorType;
  config: Record<string, unknown>;
  schedule: SyncFrequency;
  enabled: boolean;
  created_at: string;
  updated_at: string;
  last_sync_at?: string;
  health_status: HealthStatus;
}

/**
 * Sync operation details.
 */
export interface SyncOperation {
  sync_id: string;
  connector_id: string;
  status: SyncStatus;
  full_sync: boolean;
  started_at: string;
  completed_at?: string;
  records_processed?: number;
  records_failed?: number;
  error_message?: string;
  progress_percent?: number;
}

/**
 * Connection test result.
 */
export interface ConnectionTestResult {
  connection_ok: boolean;
  latency_ms: number;
  error_message?: string;
  tested_at: string;
}

/**
 * Connector health details.
 */
export interface ConnectorHealth {
  connector_id: string;
  status: HealthStatus;
  last_sync_at?: string;
  last_sync_status?: SyncStatus;
  error_count_24h: number;
  avg_sync_duration_ms?: number;
  last_error?: string;
  checked_at: string;
}

/**
 * Pagination info.
 */
export interface PaginationInfo {
  limit: number;
  offset: number;
  total: number;
  has_more: boolean;
}

/**
 * Client interface for connectors operations.
 */
interface ConnectorsClientInterface {
  request<T = unknown>(
    method: string,
    path: string,
    options?: { params?: Record<string, unknown>; json?: Record<string, unknown> }
  ): Promise<T>;
}

/**
 * Connectors API for managing enterprise data source integrations.
 *
 * @example
 * ```typescript
 * const client = createClient({ baseUrl: 'https://api.aragora.ai' });
 *
 * // List all connectors
 * const { connectors } = await client.connectors.list();
 *
 * // Create a new connector
 * const connector = await client.connectors.create(
 *   'my-postgres',
 *   'postgresql',
 *   { host: 'db.example.com', database: 'mydb' }
 * );
 *
 * // Trigger a sync
 * const sync = await client.connectors.triggerSync(connector.id);
 * ```
 */
export class ConnectorsAPI {
  constructor(private client: ConnectorsClientInterface) {}

  // ===========================================================================
  // Connector Management
  // ===========================================================================

  /**
   * List configured connectors with filtering and pagination.
   *
   * @param options - List options
   * @param options.connectorType - Filter by connector type
   * @param options.limit - Maximum number of results (default: 50)
   * @param options.offset - Pagination offset (default: 0)
   */
  async list(options?: {
    connectorType?: ConnectorType;
    limit?: number;
    offset?: number;
  }): Promise<{
    connectors: Connector[];
    pagination: PaginationInfo;
  }> {
    const params: Record<string, unknown> = {
      limit: options?.limit ?? 50,
      offset: options?.offset ?? 0,
    };
    if (options?.connectorType) {
      params.type = options.connectorType;
    }

    return this.client.request('GET', '/api/v1/connectors', { params });
  }

  /**
   * Get a connector by ID.
   *
   * @param connectorId - Connector ID
   */
  async get(connectorId: string): Promise<Connector> {
    return this.client.request('GET', `/api/v1/connectors/${connectorId}`);
  }

  /**
   * Create a new connector.
   *
   * @param name - Connector name
   * @param connectorType - Type of connector
   * @param config - Connector-specific configuration
   * @param schedule - Sync frequency (default: 'daily')
   */
  async create(
    name: string,
    connectorType: ConnectorType,
    config: Record<string, unknown>,
    schedule: SyncFrequency = 'daily'
  ): Promise<Connector> {
    return this.client.request('POST', '/api/v1/connectors', {
      json: {
        name,
        type: connectorType,
        config,
        schedule,
      },
    });
  }

  /**
   * Update a connector.
   *
   * @param connectorId - Connector ID
   * @param options - Update options
   * @param options.name - New name
   * @param options.config - New configuration
   * @param options.schedule - New sync frequency
   * @param options.enabled - Enable/disable connector
   */
  async update(
    connectorId: string,
    options?: {
      name?: string;
      config?: Record<string, unknown>;
      schedule?: SyncFrequency;
      enabled?: boolean;
    }
  ): Promise<Connector> {
    const data: Record<string, unknown> = {};
    if (options?.name !== undefined) {
      data.name = options.name;
    }
    if (options?.config !== undefined) {
      data.config = options.config;
    }
    if (options?.schedule !== undefined) {
      data.schedule = options.schedule;
    }
    if (options?.enabled !== undefined) {
      data.enabled = options.enabled;
    }

    return this.client.request('PATCH', `/api/v1/connectors/${connectorId}`, { json: data });
  }

  /**
   * Delete a connector.
   *
   * @param connectorId - Connector ID
   */
  async delete(connectorId: string): Promise<{ deleted: boolean; message: string }> {
    return this.client.request('DELETE', `/api/v1/connectors/${connectorId}`);
  }

  // ===========================================================================
  // Sync Operations
  // ===========================================================================

  /**
   * Trigger a sync for a connector.
   *
   * @param connectorId - Connector ID
   * @param fullSync - Whether to do a full sync instead of incremental (default: false)
   */
  async triggerSync(
    connectorId: string,
    fullSync = false
  ): Promise<{ sync_id: string; status: SyncStatus; message: string }> {
    return this.client.request('POST', `/api/v1/connectors/${connectorId}/sync`, {
      json: { full_sync: fullSync },
    });
  }

  /**
   * Get status of a sync operation.
   *
   * @param connectorId - Connector ID
   * @param syncId - Sync operation ID
   */
  async getSyncStatus(connectorId: string, syncId: string): Promise<SyncOperation> {
    return this.client.request('GET', `/api/v1/connectors/${connectorId}/syncs/${syncId}`);
  }

  /**
   * List recent sync operations for a connector.
   *
   * @param connectorId - Connector ID
   * @param limit - Maximum number of results (default: 20)
   */
  async listSyncs(
    connectorId: string,
    limit = 20
  ): Promise<{ syncs: SyncOperation[]; total: number }> {
    return this.client.request('GET', `/api/v1/connectors/${connectorId}/syncs`, {
      params: { limit },
    });
  }

  /**
   * Cancel a running sync operation.
   *
   * @param connectorId - Connector ID
   * @param syncId - Sync operation ID
   */
  async cancelSync(
    connectorId: string,
    syncId: string
  ): Promise<{ cancelled: boolean; message: string }> {
    return this.client.request('POST', `/api/v1/connectors/${connectorId}/syncs/${syncId}/cancel`);
  }

  // ===========================================================================
  // Health and Monitoring
  // ===========================================================================

  /**
   * Test connectivity for a connector.
   *
   * @param connectorId - Connector ID
   */
  async testConnection(connectorId: string): Promise<ConnectionTestResult> {
    return this.client.request('POST', `/api/v1/connectors/${connectorId}/test`);
  }

  /**
   * Get health status of a connector.
   *
   * @param connectorId - Connector ID
   */
  async getHealth(connectorId: string): Promise<ConnectorHealth> {
    return this.client.request('GET', `/api/v1/connectors/${connectorId}/health`);
  }
}
